sct-commons/src/main/java/org/lfenergy/compas/sct/commons/SclService.java

```diff
@@ -103,6 +103,57 @@ public void addSubnetworks(SCL scd, List<SubNetworkDTO> subNetworks, SCL icd) th
+                                .flatMap(tied -> tied.getAccessPoint().stream())
+                                .noneMatch(tAccessPoint -> tAccessPoint.getName().equals(stdConnectedAP.getApName()))) {
+                            throw new ScdException("Unknown AccessPoint :" + stdConnectedAP.getApName() + " in IED :" + iedName);
+                        }
+                        // add SubNetwork if not exist
+                        if(scd.getCommunication() == null) {
+                            scd.setCommunication(new TCommunication());
+                        }
+                        TSubNetwork tSubNetwork = scd.getCommunication().getSubNetwork().stream()
+                                .filter(subNetwork -> subNetwork.getName().equals(stdSubNetwork.getName()))
+                                .findFirst()
+                                .orElseGet(() -> {
+                                    TSubNetwork newSubNetwork = new TSubNetwork();
+                                    newSubNetwork.setName(stdSubNetwork.getName());
+                                    newSubNetwork.setType(stdSubNetwork.getType());
+                                    scd.getCommunication().getSubNetwork().add(newSubNetwork);
+                                    return newSubNetwork;
+                                });
+                        // add ConnectedAP to SubNetwork if not exist
+                        String apName = stdConnectedAP.getApName();
+                        TConnectedAP tConnectedAP = tSubNetwork.getConnectedAP().stream()
+                                .filter(connectedAP -> Objects.equals(connectedAP.getIedName(), iedName)
+                                        && Objects.equals(connectedAP.getApName(), apName))
+                                .findFirst()
+                                .orElseGet(() -> {
+                                    TConnectedAP newConnectedAP = new TConnectedAP();
+                                    newConnectedAP.setIedName(iedName);
+                                    newConnectedAP.setApName(apName);
+                                    tSubNetwork.getConnectedAP().add(newConnectedAP);
+                                    return newConnectedAP;
+                                });
+                        //copy Address And PhysConn From Icd to Scd
+                        std.getCommunication().getSubNetwork().stream()
+                                .flatMap(subNetwork -> subNetwork.getConnectedAP().stream())
+                                .filter(connectedAP -> connectedAP.getApName().equals(tConnectedAP.getApName()))
+                                .findFirst()
+                                .ifPresent(connectedAP -> {
+                                    Optional.ofNullable(connectedAP.getAddress()).ifPresent(tConnectedAP::setAddress);
+                                    tConnectedAP.getPhysConn().addAll(connectedAP.getPhysConn());
+                                });
```

Comment:
You should either remove the code in the adapter or use the Adapter. Here there is duplicated code with `connectedAPAdapter.copyAddressAndPhysConnFromIcd`.

Same thing for `communicationAdapter.addSubnetwork`

Reply:
Done I used the existing method